File: 7-1-3.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[][] D = new int[n + 1][10];
        for (int i = 0; i <= n; i++) {
            Arrays.fill(D[i], 0);
        }
        for (int j = 1; j <= 9; j++) {
            D[1][j] = 1;
        }
        for (int i = 2; i <= n; i++) {
            for (int j = 1; j <= 9; j++) {
                int s = Math.max(j - 2, 1);
                int e = Math.min(j + 2, 9);
                for (int k = s; k <= e; k++) {
                    D[i][j] += D[i - 1][k];
                    D[i][j] %= 987654321;
                }
            }
        }
        int answer = 0;
        for (int j = 1; j <= 9; j++) {
            answer += D[n][j];
            answer %= 987654321;
        }
        System.out.println(answer);
    }
}

File: 7-1-3.py
import sys
input=sys.stdin.readline

def solution(n):
    D = list([0] * 10 for _ in range(n + 1))
    for j in range(1, 10):
        D[1][j] = 1

    for i in range(2, n + 1):
        for j in range(1, 10):
            s = max(j - 2, 1)
            e = min(j + 2, 9)

            for k in range(s, e + 1):
                D[i][j] += D[i - 1][k]
            D[i][j] %= 987654321

    answer = 0
    for j in range(1, 10):
        answer += D[n][j]
    return answer % 987654321

n = int(input())
print(solution(n))